parlai/tasks/igc/agents.py

```diff
@@ -0,0 +1,336 @@
+                'https://www.microsoft.com/en-us/download/details.aspx?id=55324. '
+                'Then, make sure to put the two .csv files in ParlAI/data/igc/'
+            )
+        if (
+            not os.path.exists(self.get_image_path(opt))
+            or len(os.listdir(self.get_image_path(opt))) <= 1
+        ):
+            self._download_images(opt)
+
+        self.data = []
+        with open(dp, newline='\n') as csv_file:
+            reader = csv.reader(csv_file, delimiter=',')
+            fields = []
+            for i, row in enumerate(reader):
+                if i == 0:
+                    fields = row
+                else:
+                    ep = dict(zip(fields, row))
+                    ep['image_id'] = f'{ep["id"]}'
+                    self.data.append(ep)
+
+        if dt == 'train':
+            # Take first 90% of valid set as train
+            self.data = self.data[: int(len(self.data) * 0.9)]
+        elif dt == 'valid':
+            self.data = self.data[int(len(self.data) * 0.9) :]
+
+        self.valid_image_ids = []
+        for d in self.data:
+            img_path = os.path.join(self.get_image_path(opt), d['image_id'])
+            if os.path.isfile(img_path):
+                self.valid_image_ids.append(d['image_id'])
+
+        self.valid_image_ids = set(self.valid_image_ids)
+        return self.data
+
+    def _download_images(self, opt: Dict[str, Any]):
+        """
+        Download available IGC images.
+        """
+        urls = []
+        ids = []
+        for dt in ['test', 'val']:
+            df = os.path.join(self.get_data_path(opt), f'IGC_crowd_{dt}.csv')
+            with open(df, newline='\n') as csv_file:
+                reader = csv.reader(csv_file, delimiter=',')
+                fields = []
+                for i, row in enumerate(reader):
+                    if i == 0:
+                        fields = row
+                    else:
+                        data = dict(zip(fields, row))
+                        urls.append(data['url'])
+                        ids.append(data['id'])
+        os.makedirs(self.get_image_path(opt), exist_ok=True)
+        # Make one blank image
+        image = Image.new('RGB', (100, 100), color=0)
+        image.save(os.path.join(self.get_image_path(opt), self.blank_image_id), 'JPEG')
+        # Download the rest
+        download_multiprocess(urls, self.get_image_path(opt), dest_filenames=ids)
+
+        # Remove bad images
+        for fp in os.listdir(self.get_image_path(opt)):
+            img_path = os.path.join(self.get_image_path(opt), fp)
+            if os.path.isfile(img_path):
+                try:
+                    Image.open(img_path).convert('RGB')
+                except OSError:
+                    os.remove(img_path)
+
+    def share(self) -> Dict[str, Any]:
+        shared = super().share()
+        shared['valid_image_ids'] = self.valid_image_ids
+        return shared
+
+
+class IGCOneSideTeacher(ABC, IGCTeacher):
```

Comment:
why ABC here?

Reply:
i don't want anyone instantiating this teacher directly, as one should instead use a subclass (response only, question only)